src/metaopt/io/database/mongodb.py

```diff
@@ -0,0 +1,137 @@
+import json
+import bson
+import pymongo
+
+from pymongo import MongoClient
+from datetime import datetime
+from abstractdb import AbstractDB
+
+try:
+    from StringIO import StringIO
+except ImportError:
+    from io import StringIO
+
+try:
+    # Python 3.x
+    from urllib.parse import quote_plus
+except ImportError:
+    # Python 2.x
+    from urllib import quote_plus
+
+
+class MongoDB(AbstractDB):
+    def __init__(self, dbhost='localhost', dbname='metaopt', dbport=27017, \
+                                            username=None, password=None):
+        self.dbhost = str(dbhost)
+        self.dbname = str(dbname)
+        self.username = str(username)
+        self.password = str(password)
+
+        # we might have a different port for the database port
+        # so here we do a check if the port is there
+
+        self.dbport = str(dbport)
+
+        # set the connection string
+        # this should be mongodb://mongodb0.example.net:27017
+        # according to MongoDB documentation
+        # https://docs.mongodb.com/getting-started/python/client/
+
+        self.uri = 'mongodb://' + self.dbhost + ':' + self.dbport
+        #self.uri = "mongodb://%s:%s@%s:%s" % (quote_plus(self.username), \
+        #                                      quote_plus(self.password), \
+        #                                     self.dbhost, self.dbport)
+        # try to initialize connection
+        # setting this as a function allows to make sure reconnection
+        # is properly done in the case of an unreliable network
+
+        self.conn = self.initiate_connection()
+        self.db = self.conn[self.dbname]
+
+    def initiate_connection(self):
+        try:
+            conn = MongoClient(self.uri)
+            self.conn = conn
+        except pymongo.errors.ConnectionFailure:
+            raise Exception("Could not connect to server")
+        # return connection string for checking purposes
+        return conn
+
+    def check_connection(self):
+        if self.db:
+            #print("Connection Successful!")
+            return True
+        else:
+            #print("Please check your connection")
+            return False
+
+    def close_connection(self):
+        try:
+            close = self.conn.close()
+        except pymongo.errors.ConnectionFailure:
+            raise Exception("Could not close socket connection")
+        return close
+
+    def is_locked(self):
+        try:
+            is_locked = self.conn.is_locked()
+        except pymongo.errors.ConnectionFailure:
+            raise Exception("Could not determine if connection is locked")
+        return is_locked
+
+    def unlock(self):
+        try:
+            unlock = self.conn.unlock()
+        except pymongo.errors.ConnectionFailure:
+            raise Exception("Could determine unlock due to connection")
+
+    def insert(self, experiment_name, trial_field, dictionary, field_filters=None):
```

Comment:
trial_field is used as a collection name. It's confusing. Is trial_field supposed to be a string inside the collection `experiment_name` pointing to a dictionary of trials ({id: trial})? Shouldn't that trial_field be a `CONSTANT` variable then? 

Nevertheless, I'm not sure DB classes should have experiment and trial semantic tied to them. From Christos' UMLs, it seems to me that Experiment and Trial should have their own interface, with each using AbstractDB internally (which can be MongoDB or PostgreSQL, etc). In some way, workers and producers do not need to *know* they are communicating with a DB, anything could be hidden behind Experiment and Trial objects; Database, Sockets, whatever. @tsirif Any thoughts on that?